test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.cc

```diff
@@ -0,0 +1,303 @@
+      ENVOY_LOG_MISC(trace, "Sending chunk of {} bytes", data_size);
+      codec_client_->sendData(encoder, data_size, false);
+      fh->immediate_resp_lock_.unlock();
+    }
+
+    // See comment above
+    fh->immediate_resp_lock_.lock();
+    if (!fh->immediate_resp_sent_) {
+      ENVOY_LOG_MISC(trace, "Sending empty chunk to close stream");
+      Buffer::OwnedImpl empty_chunk;
+      codec_client_->sendData(encoder, empty_chunk, true);
+    }
+    fh->immediate_resp_lock_.unlock();
+    return response;
+  }
+
+  IntegrationStreamDecoderPtr randomDownstreamRequest(FuzzedDataProvider* fdp,
+                                                      ExtProcFuzzHelper* fh) {
+    // From the external processor's view each of these requests
+    // are handled the same way. They only differ in what the server should
+    // send back to the client.
+    // TODO(ikepolinsky): add random flag for sending trailers with a request
+    //   using HttpIntegration::sendTrailers()
+    switch (fdp->ConsumeEnum<HttpMethod>()) {
+    case HttpMethod::GET:
+      ENVOY_LOG_MISC(trace, "Sending GET request");
+      return sendDownstreamRequest(absl::nullopt);
+    case HttpMethod::POST:
+      if (fdp->ConsumeBool()) {
+        ENVOY_LOG_MISC(trace, "Sending POST request with body");
+        uint32_t data_size = fdp->ConsumeIntegralInRange<uint32_t>(0, ext_proc_fuzz_max_data_size);
+        std::string data = std::string(data_size, 'a');
+        return sendDownstreamRequestWithBody(data, absl::nullopt);
+      } else {
+        ENVOY_LOG_MISC(trace, "Sending POST request with chunked body");
+        return sendDownstreamRequestWithChunks(fdp, fh, absl::nullopt);
+      }
+    default:
+      ENVOY_LOG_MISC(error, "Unhandled HttpMethod");
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  envoy::extensions::filters::http::ext_proc::v3alpha::ExternalProcessor proto_config_{};
+  TestProcessor test_processor_;
+  Network::Address::IpVersion ip_version_;
+  Grpc::ClientType client_type_;
+};
+
+DEFINE_FUZZER(const uint8_t* buf, size_t len) {
+  // Split the buffer into two buffers with at least 1 byte
+  if (len < 2) {
+    return;
+  }
+
+  // External Process and downstream are on different threads so they should
+  // have separate data providers
+  size_t downstream_buf_len = len / 2;
+  size_t ext_proc_buf_len = len - downstream_buf_len;
+
+  // downstream buf starts at 0, ext_prob buf starts at buf[downstream_buf_len]
+  FuzzedDataProvider downstream_provider(buf, downstream_buf_len);
+  FuzzedDataProvider ext_proc_provider(&buf[downstream_buf_len], ext_proc_buf_len);
+
+  ExtProcIntegrationFuzz fuzzer(Network::Address::IpVersion::v4, Grpc::ClientType::GoogleGrpc);
+  ExtProcFuzzHelper fuzz_helper(&ext_proc_provider);
+
+  // This starts an external processor in a separate thread. This allows for the
+  // external process to consume messages in a loop without blocking the fuzz
+  // target from receiving the response.
+  fuzzer.test_processor_.start(
```

Comment:
I would be willing to follow up to see if an injected mock gRPC client improves the speed and avoids the multi-threaded approach. However, it may take some time.

Reply:
As discussed offline, let's leave some comments in this PR and punt to later.

Reply:
Added to TODO